### DISCUSSION TITLE
How to use the storage db relation.

### DISCUSSION THREAD

**Discussion by devon2018** [supabase/supabase#3541] [Questions]
I've setup my database to use a relation field to a file upload as follows:

<img width="667" alt="Screenshot 2021-10-14 at 23 18 48" src="https://user-images.githubusercontent.com/46196252/137403228-3e42e3bf-f19a-414e-a4af-8cfde10ac61a.png">

But I'm having trouble creating the relation, I assume i need to update the record to store the file uuid but have no idea and cant seem to find any docs on how to get the file uuid after upload.

**Comment by dshukertjr** [comment]
Hi @devon2018 !

What is the purpose of setting this relation? There is not a whole lot you can do with file uuids, so depending on the use case, there may be other ways to achieve what you are trying to achieve!

**Reply by devon2018** [reply]
I have a companies table and trying to store a logo for the company, I realise theres probably other ways to do this, for example using the file path or even creating a new files table and creating the link between them.

I assumed by using the file relation as above, it would have more benefits such as foreign keys, or even being able to query off the company model to get the file rather that 2 separate calls for each.

**Reply by dshukertjr** [reply]
I would like to hear how other people do it as well, but when I store files, I typically just get the file URL and store that directly on the database, in your case `companies` database. This way, all you need to do is query the `companies` database to get all the images. I know that this might be slightly lazy implementation, but it gets the job done in most cases, especially if all files are public.

**Reply by adrinr** [reply]
I am facing the same issues. I have user profiles with avatars. Not being able to have the foreign key relation makes everything flakier, as I need to store URLs instead of the proper foreign keys

**Comment by GaryAustin1** [comment]
I store the path with my notes record as my files are not public.  There is very little you can do with the storage object table so why bother with an extra table load.  I'm not sure you are even guaranteed the structure of the storage tables won't change in the future.

**Comment by mhsnook** [comment]
@devon2018 I came here with the same question! I had assumed that I would be able to add a "thumbnail" column to my "posts" table and be able to do something like post.thumbnail and view the thumbnail. Or maybe have some relation like `post.thumbnail.public_url` or `post.thumbnail_public_url`.

But it seems that what I'm supposed to do is, not submit my image _with_ my post, but submit it separately via the storage API, and then get the URL from the response, and then add the URL as a hidden input in my post creation form. But this is difficult to work out because the return values for the storage API are not really documented!

Anyway, it seems like you have to treat the database and the storage database as totally separate entities and manage all the connections via plaintext fields and JS linkages, which is... strange, to say the least.

**Reply by nicogs** [reply]
strange that this isn't possible (yet?)

**Comment by intagaming** [comment]
Suppose that I am in the same boot as the OP. In the case of me not being able to use the object's uuid as a Foreign Key and only storing the object's name, if I delete the object, then there's nothing I can do to update my row (at least conveniently and automatically through the `ON DELETE CASCADE`). There's just no relationship, but indeed there is.

Why not just return the `FileObject` when I do `supabase.storage.from().upload()`? I could get the hold of `FileObject[]` through the `supabase.storage.from().list()` and filter the name out to get the object's uuid if I want. But why just return `{Key: string}`? What is `Key` with the capital `K` anyway?

**Reply by avalanche-tm** [reply]
Is there any update on this? Seems like it's still impossible to do relations between storage.objects and other tables. This is such a bummer!

**Comment by bwalsh1191** [comment]
Coming here for hopefully some updates!

**Comment by caseycrogers** [comment]
Same! It'd be really cool if I could enforce some sort of foreign key constraint across a postgres table and a storage bucket s.t. I could store a URL in the table and guarantee that it's a valid reference to a storage file in the correct bucket.

**Comment by david-plugge** [comment]
Since you need the bucket id AND the object name you have to create a foreign key with 2 columns:

```sql
create table public.users (
  id bigint generated by default as identity not null,
  created_at timestamp with time zone null default now(),
  name text not null,
  image_bucket_id text null,
  image_name text null,
  
  constraint ingredients_pkey primary key (id),
  -- add this foreign key
  foreign key (image_bucket_id, image_name) references storage.objects ( bucket_id, name ) on delete set null
);

-- or add the constraint after creating the table using the dashboard
alter table public.users add constraint fk_users_image
  foreign key (image_bucket_id, image_name)
  references storage.objects ( bucket_id, name ) on delete set null;

```

Using the js sdk to get the public url:

```ts
const { image_bucket_id, image_name, ...restUser } = await supabaseClient
  .from('users')
  .select()
  .eq('email', email)
  .single();

const userImage = supabaseClient
  .storage
  .from(image_bucket_id)
  .getPublicUrl(image_name);

const user = {
  image: userImage,
  ...restUser
}
```

It does work, but it´s still a little annoying that we can´t create the relation in the dashboard.

**Reply by AlexanderBuzzCrafter** [reply]
To my understanding this only works with individual images, not an array of images.

**Comment by Droces** [comment]
Doing what @david-plugge suggested now works in the Supabase UI 😀 It lets you create a primary key based on multiple columns.

**Comment by ypotsiah** [comment]
Some questions to Supabase guys:
1. Is it safe to set FK relation to storage.objects PK from own tables? Seems it's not guaranteed to have such DB schema unchanged by [Supabase Docs](https://supabase.com/docs/guides/auth/managing-user-data)
2. Is it enough to delete row from storage.objects to remove file? Do you have any kind of sync logic between DB and S3 bucket ?

**Comment by GaryAustin1** [comment]
1) It is safe, not sure how useful as you need path to do any API call and the REST API can't join between schemas.
2) Absolutely NOT.  You will orphan your file data in S3.
